docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,276 @@
+
+Following this first strategy, you do the following tasks:
+
+1. Enable nullable types for the entire project by adding the `<Nullable>enable</Nullable>` element to your *csproj* files. 
+1. Add the `#nullable disable` pragma to every source file in your project. 
+1. As you work on each file, remove the pragma and address any warnings.
+
+This first strategy has more up-front work to add the pragma to every file. The advantage is that every new code file added to the project will be nullable enabled. Any new work will be nullable aware; only existing code must be updated.
+
+The second strategy works better if the library is generally stable, and the main focus of the development is to adopt nullable reference types. You turn on nullable reference types as you annotate APIs. When you've finished, you enable nullable reference types for the entire project.
+
+Following this second strategy you do the following tasks:
+
+1. Add the `#nullable enable` pragma to the file you want to make nullable aware.
+1. Address any warnings.
+1. Continue these first two steps until you have made the entire library nullable aware.
+1. Enable nullable types for the entire project by adding the `<Nullable>enable</Nullable>` element to your *csproj* files. 
+1. Remove the `#nullable enable` pragmas, as they're no longer needed.
+
+This second strategy has less work up-front. The tradeoff is that the first task when you create a new file is to add the pragma and make it nullable aware. If any developers on your team forget, that new code is now in the backlog of work to make all code nullable aware.
+
+Which of these strategies you pick depends on how much active development is taking place in your project. The more mature and stable your project, the better the second strategy. The more active features are being developed, the better the first strategy.
+
+## Should nullable warnings introduce breaking changes?
+
+Before you enable nullable reference types, variables are considered *nullable oblivious*. Once you enable nullable reference types, all those variables are *non-nullable*. The compiler will issue warnings if those variables aren't initialized to non-null values.
+
+Another likely source of warnings is return values when the value hasn't been initialized.
+
+The first step is to use `?` annotations on parameters and return types to indicate when arguments or return values may be null, or must not be null. As you do this, your goal isn't just to fix warnings. The more important goal is to make the compiler understand your intent for potential null values. As you examine the warnings, you reach your next major decision for your library. Do you want to consider modifying API signatures to more clearly communicate your design intent? A better API signature for the `TryGetMessage` method examined earlier could be:
```

Comment:
```suggestion
The first step in addressing the compiler warnings is to use `?` annotations on parameters and return types to indicate when arguments or return values may be null, or must not be null. As you do this, your goal isn't just to fix warnings. The more important goal is to make the compiler understand your intent for potential null values. As you examine the warnings, you reach your next major decision for your library. Do you want to consider modifying API signatures to more clearly communicate your design intent? A better API signature for the `TryGetMessage` method examined earlier could be:
```
I removed the comma, which doesn't belong if "or must not be null" applies to using ? annotations, though I don't think that's really the correct change. It is if the ? annotation indicates that a parameter or return type must not be null -- but a nullable type indicates that it may be null.

Reply:
I split this into two sentences to be more clear.